File: backend/app/models/shop.py
from typing import Optional, List
from pydantic import BaseModel, field_validator


class ShopCreate(BaseModel):
    shop_name: str
    owner_name: str
    mobile: str
    address: Optional[str] = None
    initial_udhari: Optional[float] = 0.0


    @field_validator("shop_name", "owner_name", "mobile")
    @classmethod
    def validate_non_empty(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("Field cannot be empty")
        return v.strip()


class ShopUpdate(BaseModel):
    shop_name: Optional[str] = None
    owner_name: Optional[str] = None
    mobile: Optional[str] = None
    address: Optional[str] = None


class ShopResponse(BaseModel):
    id: str
    shop_name: str
    owner_name: str
    mobile: str
    address: Optional[str]
    is_archived: bool
    created_at: str
    updated_at: Optional[str] = None
    total_purchases: Optional[int] = 0
    pending_udhari: Optional[float] = 0.0

    @classmethod
    def from_row(cls, row: dict) -> "ShopResponse":
        return cls(
            id=str(row["id"]),
            shop_name=row["shop_name"],
            owner_name=row["owner_name"],
            mobile=row["mobile"],
            address=row.get("address"),
            is_archived=row.get("is_archived", False),
            created_at=str(row.get("created_at", "")),
            updated_at=str(row["updated_at"]) if row.get("updated_at") else None,
            total_purchases=int(row.get("total_purchases", 0)),
            pending_udhari=float(row.get("pending_udhari", 0.0)),
        )


class ShopPurchaseCreate(BaseModel):
    battery_model: str
    serial_number: str
    invoice_number: str = ""
    quantity: int = 1
    purchase_date: str
    amount: float
    udhari_amount: float = 0.0

    @field_validator("battery_model", "serial_number")
    @classmethod
    def validate_non_empty(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("Field cannot be empty")
        return v.strip()

    @field_validator("quantity")
    @classmethod
    def validate_qty(cls, v: int) -> int:
        if v <= 0:
            raise ValueError("quantity must be greater than 0")
        return v

    @field_validator("amount")
    @classmethod
    def validate_amount(cls, v: float) -> float:
        if v < 0:
            raise ValueError("amount cannot be negative")
        return round(v, 2)

    @field_validator("udhari_amount")
    @classmethod
    def validate_udhari(cls, v: float) -> float:
        if v < 0:
            raise ValueError("udhari_amount cannot be negative")
        return round(v, 2)


class ShopPurchaseResponse(BaseModel):
    id: str
    shop_id: str
    battery_model: str
    serial_number: str
    invoice_number: str
    quantity: int
    purchase_date: str
    amount: float
    udhari_amount: float
    created_at: str

    @classmethod
    def from_row(cls, row: dict) -> "ShopPurchaseResponse":
        return cls(
            id=str(row["id"]),
            shop_id=str(row["shop_id"]),
            battery_model=row["battery_model"],
            serial_number=row["serial_number"],
            invoice_number=row["invoice_number"],
            quantity=int(row.get("quantity", 1)),
            purchase_date=str(row.get("purchase_date", "")),
            amount=float(row.get("amount", 0.0)),
            udhari_amount=float(row.get("udhari_amount", 0.0)),
            created_at=str(row.get("created_at", "")),
        )


class ShopPaymentResponse(BaseModel):
    id: str
    shop_id: str
    total_amount: float
    paid_amount: float
    pending_amount: float
    is_settled: bool
    created_at: str

    @classmethod
    def from_row(cls, row: dict) -> "ShopPaymentResponse":
        return cls(
            id=str(row["id"]),
            shop_id=str(row["shop_id"]),
            total_amount=float(row.get("total_amount", 0.0)),
            paid_amount=float(row.get("paid_amount", 0.0)),
            pending_amount=float(row.get("pending_amount", 0.0)),
            is_settled=row.get("is_settled", False),
            created_at=str(row.get("created_at", "")),
        )


class ShopPaymentTransactionResponse(BaseModel):
    id: str
    payment_id: str
    shop_id: str
    transaction_type: str
    amount: float
    notes: Optional[str]
    created_at: str

    @classmethod
    def from_row(cls, row: dict) -> "ShopPaymentTransactionResponse":
        return cls(
            id=str(row["id"]),
            payment_id=str(row["payment_id"]),
            shop_id=str(row["shop_id"]),
            transaction_type=row["transaction_type"],
            amount=float(row.get("amount", 0.0)),
            notes=row.get("notes"),
            created_at=str(row.get("created_at", "")),
        )


class ShopDetailsResponse(BaseModel):
    shop: ShopResponse
    purchases: List[ShopPurchaseResponse]
    payment: Optional[ShopPaymentResponse] = None
    transactions: List[ShopPaymentTransactionResponse] = []

File: frontend/lib/models/shop.dart
class Shop {
  final String id;
  final String shopName;
  final String ownerName;
  final String mobile;
  final String? address;
  final bool isArchived;
  final String createdAt;
  final String? updatedAt;
  final int totalPurchases;
  final double pendingUdhari;

  Shop({
    required this.id,
    required this.shopName,
    required this.ownerName,
    required this.mobile,
    this.address,
    required this.isArchived,
    required this.createdAt,
    this.updatedAt,
    this.totalPurchases = 0,
    this.pendingUdhari = 0.0,
  });

  factory Shop.fromJson(Map<String, dynamic> json) {
    return Shop(
      id: json['id'] as String,
      shopName: json['shop_name'] as String,
      ownerName: json['owner_name'] as String,
      mobile: json['mobile'] as String,
      address: json['address'] as String?,
      isArchived: json['is_archived'] as bool? ?? false,
      createdAt: json['created_at'] as String,
      updatedAt: json['updated_at'] as String?,
      totalPurchases: json['total_purchases'] as int? ?? 0,
      pendingUdhari: (json['pending_udhari'] as num? ?? 0.0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'shop_name': shopName,
      'owner_name': ownerName,
      'mobile': mobile,
      'address': address,
      'is_archived': isArchived,
      'created_at': createdAt,
      'updated_at': updatedAt,
      'total_purchases': totalPurchases,
      'pending_udhari': pendingUdhari,
    };
  }
}

class ShopPurchase {
  final String id;
  final String shopId;
  final String batteryModel;
  final String serialNumber;
  final String invoiceNumber;
  final int quantity;
  final String purchaseDate;
  final double amount;
  final double udhariAmount;
  final String createdAt;
  final String? paymentMode;

  ShopPurchase({
    required this.id,
    required this.shopId,
    required this.batteryModel,
    required this.serialNumber,
    required this.invoiceNumber,
    required this.quantity,
    required this.purchaseDate,
    required this.amount,
    required this.udhariAmount,
    required this.createdAt,
    this.paymentMode,
  });

  factory ShopPurchase.fromJson(Map<String, dynamic> json) {
    return ShopPurchase(
      id: json['id'] as String,
      shopId: json['shop_id'] as String,
      batteryModel: json['battery_model'] as String,
      serialNumber: json['serial_number'] as String,
      invoiceNumber: json['invoice_number'] as String,
      quantity: json['quantity'] as int? ?? 1,
      purchaseDate: json['purchase_date'] as String,
      amount: (json['amount'] as num? ?? 0.0).toDouble(),
      udhariAmount: (json['udhari_amount'] as num? ?? 0.0).toDouble(),
      createdAt: json['created_at'] as String,
      paymentMode: json['payment_mode'] as String?,
    );
  }
}

class ShopPayment {
  final String id;
  final String shopId;
  final double totalAmount;
  final double paidAmount;
  final double pendingAmount;
  final bool isSettled;
  final String createdAt;

  ShopPayment({
    required this.id,
    required this.shopId,
    required this.totalAmount,
    required this.paidAmount,
    required this.pendingAmount,
    required this.isSettled,
    required this.createdAt,
  });

  factory ShopPayment.fromJson(Map<String, dynamic> json) {
    return ShopPayment(
      id: json['id'] as String,
      shopId: json['shop_id'] as String,
      totalAmount: (json['total_amount'] as num? ?? 0.0).toDouble(),
      paidAmount: (json['paid_amount'] as num? ?? 0.0).toDouble(),
      pendingAmount: (json['pending_amount'] as num? ?? 0.0).toDouble(),
      isSettled: json['is_settled'] as bool? ?? false,
      createdAt: json['created_at'] as String,
    );
  }
}

class ShopPaymentTransaction {
  final String id;
  final String paymentId;
  final String shopId;
  final String transactionType;
  final double amount;
  final String? notes;
  final String createdAt;
  final String? paymentMode;

  ShopPaymentTransaction({
    required this.id,
    required this.paymentId,
    required this.shopId,
    required this.transactionType,
    required this.amount,
    this.notes,
    required this.createdAt,
    this.paymentMode,
  });

  factory ShopPaymentTransaction.fromJson(Map<String, dynamic> json) {
    return ShopPaymentTransaction(
      id: json['id'] as String,
      paymentId: json['payment_id'] as String,
      shopId: json['shop_id'] as String,
      transactionType: json['transaction_type'] as String,
      amount: (json['amount'] as num? ?? 0.0).toDouble(),
      notes: json['notes'] as String?,
      createdAt: json['created_at'] as String,
      paymentMode: json['payment_mode'] as String?,
    );
  }
}

class ShopDetails {
  final Shop shop;
  final List<ShopPurchase> purchases;
  final ShopPayment? payment;
  final List<ShopPaymentTransaction> transactions;

  ShopDetails({
    required this.shop,
    required this.purchases,
    this.payment,
    required this.transactions,
  });

  factory ShopDetails.fromJson(Map<String, dynamic> json) {
    return ShopDetails(
      shop: Shop.fromJson(json['shop'] as Map<String, dynamic>),
      purchases: (json['purchases'] as List<dynamic>?)
              ?.map((e) => ShopPurchase.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      payment: json['payment'] != null
          ? ShopPayment.fromJson(json['payment'] as Map<String, dynamic>)
          : null,
      transactions: (json['transactions'] as List<dynamic>?)
              ?.map((e) => ShopPaymentTransaction.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}
